internal/engine/wazevo/frontend/frontend.go

```diff
@@ -438,12 +442,21 @@ func (c *Compiler) recordKnownSafeBound(v ssa.ValueID, safeBound uint64, absolut
 
 // clearSafeBounds clears the known safe bounds. This must be called
 // after the compilation of each block.
-func (c *Compiler) clearSafeBounds() {
-	for _, v := range c.knownSafeBoundsSet {
-		ptr := &c.knownSafeBounds[v]
-		ptr.bound = 0
+// If `hard` is true, it clears the bounds. Otherwise, it clears only the absolute addresses,
+// so that the constant bounds are still valid.
+func (c *Compiler) clearSafeBounds(hard bool) {
+	if hard {
+		for _, v := range c.knownSafeBoundsSet {
+			ptr := &c.knownSafeBounds[v]
+			ptr.bound = 0
+		}
+		c.knownSafeBoundsSet = c.knownSafeBoundsSet[:0]
+	} else {
+		for _, v := range c.knownSafeBoundsSet {
+			ptr := &c.knownSafeBounds[v]
+			ptr.absoluteAddr = ssa.ValueInvalid
+		}
```

Comment:
I would recommend splitting this method in two instead of passing a boolean, especially since it seems that we always call it with a constant true or false:

- the method body clearly shows it's doing two things, that's a good indication that it would be better suited to have the logic in two methods

- at the call site, true/false says nothing about the meaning of the literal; this can be improved if the two methods were named `clearSafeBounds` and `clearAbsoluteAddresses`

Reply:
agreed, thank you for the third eye!